### PR TITLE
Signup: Add styles for a Woo signup

### PR DESCRIPTION
**Pull request by scruffian**
This pull request adds Woo specific styles to signup:

screen shot 2017-08-24 at 18 56 08

This is just a first attempt. It will need more work once we merge 12810
  
#### Testing instructions
  
1. Run `git checkout update/oauth-signup` and start your server
2. Enable your proxy
3. Open the WooCommerce [`Home` page](https://woocommerce.com/)
4. Click the `SIGN IN WITH WORDPRESS.COM` link
5. Assert that you are redirected to the new [`Login` page](https://wordpress.com/log-in)
6. Replace `https://wordpress.com` with `http://calypso.localhost:3000` in your browser's address bar
7. Assert that you are presented with the same page, from your server this time
8. Click the `Create an Account.` link
9. Assert that you are redirected to the new dedicated [`Signup` page](http://calypso.localhost:3000/start/wpcc)
10. Assert that this page has the WooCommerce look-and-feel
11. Create a new account and assert you're redirected back to WooCommerce [`Choose Account` page](https://woocommerce.com/choose-account/)

You can also run unit tests with:

```
npm run test-client client/state/oauth2-clients
npm run test-client client/state/ui/oauth2-clients
```
#### Reviews
  
- [ ] Code
- [x] Product

### CODE REVIEW

File: client/state/login/oauth2/reducer.js [hunk 1]
```diff
@@ -115,6 +115,7 @@ export const clients = createReducer( initialClientsData, {
 
 export const currentClientId = createReducer( null, {
 	[ ROUTE_SET ]: ( state, { path, query } ) => startsWith( path, '/log-in' ) && query.client_id || null,
+	[ ROUTE_SET ]: ( state, { path, query } ) => startsWith( path, '/start' ) && query.client_id || null,
```
Comment (scruffian) [inline client/state/login/oauth2/reducer.js]: Should we move this to a new branch of the state tree?

Comment (stephanethomas) [reply client/state/login/oauth2/reducer.js]: What problem would you solve by doing so?

**Comment by stephanethomas** [comment]
A heads-up that I got stuck during the account creation process, right after clicking the `Continue` button:

File: client/state/login/oauth2/reducer.js (hunk 1 above)
Comment (stephanethomas) [reply client/state/login/oauth2/reducer.js]: The following test fails because of this change by the way:

screenshot

**Comment by stephanethomas** [comment]
I've pushed a commit to fix the [failing test](https://github.com/Automattic/wp-calypso/pull/17504/files#r135246667) I noticed earlier.

**Comment by scruffian** [comment]
> The client_id=50916 query parameter is lost when the page is refreshed, resulting in presenting the default WordPress.com look-and-feel. Entering the original url again always redirects to this default page after that.

Since we merged #17526 this no longer happens.

**Comment by scruffian** [comment]
> There is a moment when Calypso loads and displays the WordPress.com app shell. This is obviously because the page is not rendered server-side, but this creates a flash of content that will likely affect our users.

I _think_ users who come here will already have been on /log-in so they won't need a server side rendering. We can test that in 17530.

**Comment by scruffian** [comment]
> I suspect this wasn't part of the scope of this pull request, but I wanted to note that there are still a few differences between the original Woo page and the new Calypso one:

I was waiting for #17526 to be merged before working on that. I think we should handle it in a new PR.

File: client/state/login/oauth2/reducer.js (hunk 1 above)
Comment (scruffian) [reply client/state/login/oauth2/reducer.js]: Well this is in the `login` branch which doesn't make sense since this is a `signup` feature too now.

File: client/state/login/oauth2/reducer.js [hunk 2]
```diff
@@ -114,7 +114,9 @@ export const clients = createReducer( initialClientsData, {
 } );
 
 export const currentClientId = createReducer( null, {
-	[ ROUTE_SET ]: ( state, { path, query } ) => startsWith( path, '/log-in' ) && query.client_id || null,
+	[ ROUTE_SET ]: ( state, { path, query } ) => {
+		return ( startsWith( path, '/log-in' ) || startsWith( path, '/start' ) ) && query.client_id || null;
```
Comment (Tug) [inline client/state/login/oauth2/reducer.js]: The query params is called `oauth_client_id` in the case of signup.

File: client/state/ui/oauth2-clients/selectors.js [hunk 3]
```diff
@@ -10,13 +10,13 @@ import { get } from 'lodash';
...
 export const getOAuth2ClientData = ( state ) => {
-	const currentClientId = get( state, 'login.oauth2.currentClientId', null );
+	const currentClientId = get( state, 'ui.oauth2Clients.currentClientId', null );
 
 	if ( ! currentClientId ) {
 		return null;
 	}
 
-	return get( state, 'login.oauth2.clients', {} )[ currentClientId ];
+	return get( state, 'ui.oauth2Clients.clients', {} )[ currentClientId ];
```
Comment (Tug) [inline client/state/ui/oauth2-clients/selectors.js]: My suggestion would be to have `oauth2Clients` as a list at the root of the state tree. And `ui` would have everything related to the current client id, a bit like `users` and `currentUserId` at the moment.
With this organisation,
- `getOAuth2ClientData` would live in `state/ui/oauth2-clients/selectors.js`
- `showOAuth2Layout` would live in `state/ui/oauth2-clients/selectors.js` (we could probably rename it btw as it sounds like an action)
- `currentClientId` would live in `state/ui/oauth2-clients/reducer.js` like it already does
- `oauth2Clients` would live in `state/oauth2-clients/reducer.js`

File: client/blocks/login/login-form.jsx [hunk 4]
```diff
@@ -167,7 +167,7 @@ export class LoginForm extends Component {
 		if ( isOauthLogin ) {
 			signupUrl = '/start/wpcc?' + qs.stringify( { oauth2_client_id: oauth2ClientData.id, oauth2_redirect: redirectTo } );
 			// TODO remove the following line when WPCC signup is ready
-			signupUrl = oauth2ClientData.signupUrl;
+			//signupUrl = oauth2ClientData.signupUrl;
```
Comment (Tug) [inline client/blocks/login/login-form.jsx]: Memo to not forget to undo this change before we merge

**Comment by stephanethomas** [comment]
> > The client_id=50916 query parameter is lost when the page is refreshed, resulting in presenting the default WordPress.com look-and-feel. Entering the original url again always redirects to this default page after that.
> 
> Since we merged #17526 this no longer happens.

I've rebased this pull request but still encounter the same behavior.

I've also noted that accessing [this url](http://calypso.localhost:3000/start/wpcc?client_id=50916) gives the following error:

```
Error: Reducer "currentClientId" returned undefined handling "ROUTE_SET". To ignore an action, you must explicitly return the previous state.
    at /var/sources/node_modules/redux/lib/utils/combineReducers.js:114:15
    at /var/sources/node_modules/redux/lib/utils/mapValues.js:15:19
    at Array.reduce (native)
    at Object.mapValues [as default] (/var/sources/node_modules/redux/lib/utils/mapValues.js:14:27)
    at combination (/var/sources/node_modules/redux/lib/utils/combineReducers.js:109:49)
    at /var/sources/node_modules/redux/lib/utils/combineReducers.js:111:29
    at /var/sources/node_modules/redux/lib/utils/mapValues.js:15:19
    at Array.reduce (native)
    at Object.mapValues [as default] (/var/sources/node_modules/redux/lib/utils/mapValues.js:14:27)
    at combination (/var/sources/node_modules/redux/lib/utils/combineReducers.js:109:49)
    at ui (/var/sources/build/webpack:/client/state/ui/reducer.js:122:9)
    at /var/sources/node_modules/redux/lib/utils/combineReducers.js:111:29
    at /var/sources/node_modules/redux/lib/utils/mapValues.js:15:19
    at Array.reduce (native)
    at Object.mapValues [as default] (/var/sources/node_modules/redux/lib/utils/mapValues.js:14:27)
    at combination (/var/sources/node_modules/redux/lib/utils/combineReducers.js:109:49)
    at dispatch (/var/sources/node_modules/redux/lib/createStore.js:126:22)
    at /var/sources/build/webpack:/client/state/notices/middleware.js:299:9
    at /var/sources/build/webpack:/client/state/data-layer/wpcom-api-middleware.js:76:11
    at Object.dispatch (/var/sources/node_modules/redux-thunk/lib/index.js:12:74)
    at setRouteMiddleware (/var/sources/build/webpack:/server/isomorphic-routing/index.js:45:16)
    at /var/sources/build/webpack:/server/isomorphic-routing/index.js:79:67
```

I've added a commit to fix that.

**Comment by stephanethomas** [comment]
I think the testing instructions are wrong: the url to use should be http://calypso.localhost:3000/start/wpcc?oauth2_client_id=50916 instead of http://calypso.localhost:3000/start/account/user?client_id=50916.

**Comment by stephanethomas** [comment]
I've addressed @Tug's comment, fixed linting errors, and updated the testing instructions. I've also added a feature gate to ease management of the new WPCC signup flow, which is now only enabled in development.

This is good for another review.

File: client/state/ui/oauth2-clients/selectors.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	return getOAuth2Client( state, currentClientId );
+};
+
+/***
+ * Determines if the OAuth2 layout should be used.
+ *
+ * @param  {Object}   state  Global state tree
+ * @return {Boolean}         Whether the OAuth2 layout should be used.
+ */
+export const showOAuth2Layout = ( state ) => !! getCurrentOAuth2ClientId( state );
```
Comment (stephanethomas) [inline client/state/ui/oauth2-clients/selectors.js]: From what I can see, we could remove this `showOAuth2Layout` entirely instead of just renaming it. It's only used [here](https://github.com/Automattic/wp-calypso/blob/7760d4f8f00ade7b490ff66b6276744e72b16369/client/layout/logged-out.jsx#L34), where `hasValidOAuth2ClientData` is serving the same purpose.

File: client/layout/masterbar/oauth-client.scss [hunk 6]
```diff
@@ -226,7 +236,7 @@
 		font-size: 15px;
 	}
 
-	.login__form-social {
+	.login__form-social{
```
Comment (jsnmoon) [inline client/layout/masterbar/oauth-client.scss]: Minor: Probably an unintended change?

**Comment by stephanethomas** [comment]
I suggest fixing the [`Signup Processing` page](https://github.com/Automattic/wp-calypso/pull/17504#issuecomment-324903420) in another pull request.

> SSR seems to be functioning properly, with the exception of this one warning about mismatched checksum in the console:

I did not encounter that warning. Did you restart your server before testing?

**Comment by Tug** [comment]
I just tested and I encountered a weird behavior, at first I saw the loading screen as in https://github.com/Automattic/wp-calypso/pull/17504#issuecomment-324903420 then it went back to the signup screen with an error (`invalid client id` I think) then automatically redirected back to the loading screen and finally I ended up in woocommerce.
I'll investigate what's happening.

Otherwise the code LGTM 👍

File: client/layout/masterbar/oauth-client.scss [hunk 7]
```diff
@@ -211,12 +218,15 @@
 		}
 	}
 
-	.card {
-		box-shadow: none;
-		padding-left: 0;
-		padding-right: 0;
+	.step-wrapper {
+		margin-bottom: 50px;
```
Comment (Tug) [inline client/layout/masterbar/oauth-client.scss]: missing `width: 778px;` here

**Comment by Tug** [comment]
I could not reproduce...
We'll need to update the copy in the signup screen but that can be done in another PR.
Product 👍
